### PR TITLE
Add discovery diagnostics reporting for DL-PR-07

### PR DESCRIPTION
**Pull request by shaypal5**
## Summary
This PR implements the `DL-PR-07` observability slice for the discovery/candidacy layer and adds a short top-level `PLAN.md` clarifying how the repo’s planning documents relate to each other.

The core outcome is a new discovery diagnostics/reporting path that can:
- summarize engine overlap
- summarize source-native vs search-engine coverage
- report queue-health metrics
- report candidate conversion metrics
- render the report as text or JSON from the CLI
- persist a combined latest diagnostics artifact alongside the existing overlap artifact

## Planning context
This PR also adds `PLAN.md` at the repo root to make the planning structure explicit:
- `docs/CHATGPT_26_04_PLAN.md` remains the main roadmap
- `docs/MILESTONE_3_VALIDATION_PR_BREAKDOWN.md` is the Milestone 3 branch plan under that roadmap
- `docs/tfht_discovery_layer_implementation_plan.md` is the separate `DL-PR-*` architecture plan for the discovery/candidacy split

### 1. New discovery diagnostics module
Added `src/denbust/diagnostics/discovery.py` with typed models and helpers for:
- engine overlap metrics
- source-native vs search-engine coverage metrics
- queue health metrics
- candidate conversion metrics
- top scrape-failure summaries by source/domain/error code
- latest artifact persistence
- human-readable report rendering

### 2. New CLI command
Added `denbust diagnose-discovery` in `src/denbust/cli.py`.

Supported behavior:
- defaults to `agents/news/local.yaml`
- supports `--stale-days`
- supports `--format text|json`
- supports `--output` for JSON export

### 3. Discovery artifact wiring in the pipeline
Updated `run_news_discover_job` in `src/denbust/pipeline.py` so the latest discover run now writes:
- `engine_overlap_latest.json`
- `discovery_diagnostics_latest.json`

The current-run artifact generation uses the merged candidates returned by the active producer runs, so the report stays correct even when tests stub the discovery runners directly.

### 4. State-path coverage
Extended `src/denbust/discovery/state_paths.py` so the discovery state model now exposes:
- `candidate_provenance_path`
- `scrape_attempts_path`
- `discovery_diagnostics_latest_path`

This lets the diagnostics layer read the mirrored state-repo JSONL/JSON artifacts directly without depending on backend-specific persistence readers.

### 5. Tests
Added and updated unit coverage for:
- discovery diagnostics report building
- discovery diagnostics artifact persistence
- discovery state paths
- CLI wiring for `diagnose-discovery`
- discover-job artifact generation behavior

## Validation
Ran:
- `ruff check src/denbust/diagnostics/discovery.py src/denbust/diagnostics/__init__.py src/denbust/cli.py src/denbust/pipeline.py src/denbust/discovery/state_paths.py src/denbust/discovery/storage.py tests/unit/test_discovery_diagnostics.py tests/unit/test_discovery_state_paths.py tests/unit/test_cli.py tests/unit/test_pipeline_core.py`
- `mypy src`
- `pytest -q tests/unit/test_discovery_diagnostics.py tests/unit/test_discovery_state_paths.py tests/unit/test_cli.py tests/unit/test_pipeline_core.py`

All passed.

## Scope notes
This PR stays inside `DL-PR-07` scope:
- no backfill work
- no self-healing work
- no scrape-semantic changes
- no release-policy changes

It focuses strictly on observability, report generation, and developer/operator access to the new discovery-layer state.

### CODE REVIEW

File: src/denbust/pipeline.py [hunk 1]
```diff
@@ -1270,50 +1254,41 @@ async def run_news_discover_job(
...
+            merged_candidates_by_id[candidate.candidate_id] = candidate.model_copy(
+                update={
+                    "discovered_via": deduplicate_strings(
+                        [
+                            *(existing_candidate.discovered_via if existing_candidate else []),
+                            producer_name,
+                        ]
+                    )
+                }
+            )
```
[the remark this answers is not in the document]
Comment (shaypal5) [reply src/denbust/pipeline.py:1295]: Implemented with a split fix: the merged candidate state now preserves each candidate's existing `discovered_via` values instead of collapsing them to the active producer name, and the current-run overlap artifact is tracked separately from that historical provenance. This keeps durable per-producer conversion state intact without polluting the run-scoped overlap metrics.

File: src/denbust/pipeline.py [hunk 2]
```diff
@@ -1270,50 +1254,41 @@ async def run_news_discover_job(
...
 
-    _write_discovery_engine_metrics(
+    _write_discovery_diagnostic_artifacts(
         config,
-        source_native_candidate_ids=source_native_candidate_ids,
-        brave_candidate_ids=brave_candidate_ids,
-        exa_candidate_ids=exa_candidate_ids,
-        google_cse_candidate_ids=google_cse_candidate_ids,
+        config_path=config_path,
+        candidates=list(merged_candidates_by_id.values()),
```
[the remark this answers is not in the document]
Comment (shaypal5) [reply src/denbust/pipeline.py]: Updated the pipeline artifact path to build `discovery_diagnostics_latest.json` from persisted state when the durable candidate/attempt files exist, and to use the current run only for the overlap section. The override now remains only as a fallback for stubbed test paths where the persistence side effects are mocked out.

File: PLAN.md [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+## Current Next Focus: `DL-PR-07`
+
+`DL-PR-07` is the next planned discovery-layer step. Its purpose is to make the discovery/candidacy layer measurable and debuggable, building on the existing `DL-PR-01` through `DL-PR-06` foundation.
+
+### What already exists
+
+- Candidate persistence, scrape attempts, and queue state already exist under `src/denbust/discovery/`.
+- A lightweight overlap artifact already exists at `state_repo/.../metrics/engine_overlap_latest.json`.
+- The pipeline already aggregates per-engine candidate IDs in `src/denbust/pipeline.py`.
```
[the remark this answers is not in the document]
Comment (shaypal5) [reply PLAN.md]: Updated the `PLAN.md` bullet to point at `src/denbust/diagnostics/discovery.py` as the current overlap/diagnostics path.

File: src/denbust/diagnostics/discovery.py [hunk 4]
```diff
@@ -0,0 +1,570 @@
...
+    except Exception as exc:
+        notes.append(f"Operational record fetch failed: {type(exc).__name__}: {exc}")
+        return set(), notes
+    finally:
+        store.close()
+
+    urls = {
+        canonicalize_news_url(str(row["canonical_url"])) for row in rows if row.get("canonical_url")
+    }
+    return urls, notes
```
[the remark this answers is not in the document]
Comment (shaypal5) [reply src/denbust/diagnostics/discovery.py:491]: Implemented the explicit opt-out path here: pipeline-generated diagnostics artifacts now call the report builder with `include_operational_matches=False`, so discover jobs no longer fetch operational records just to refresh diagnostics. The CLI/default report path still supports full operational matching.

File: src/denbust/diagnostics/discovery.py [hunk 5]
```diff
@@ -0,0 +1,582 @@
...
+        candidates_override
+        if candidates_override is not None
+        else _read_jsonl(paths.latest_candidates_path, PersistentCandidate)
+    )
+    attempts = (
+        attempts_override
+        if attempts_override is not None
+        else _read_jsonl(paths.scrape_attempts_path, ScrapeAttempt)
+    )
+    overlap_candidates = overlap_candidates_override or candidates
```
[the remark this answers is not in the document]
Comment (shaypal5) [reply src/denbust/diagnostics/discovery.py]: Changed this to an explicit `is None` check so an empty overlap override remains a valid run-scoped signal. I also added a regression test covering the zero-candidate current-run case.

File: src/denbust/diagnostics/discovery.py [hunk 6]
```diff
@@ -0,0 +1,582 @@
...
+        ),
+        top_failure_sources=_build_failure_summaries(candidates, attempts, key="source"),
+        top_failure_domains=_build_failure_summaries(candidates, attempts, key="domain"),
+        top_failure_codes=_build_failure_summaries(candidates, attempts, key="error_code"),
+    )
+    if not candidates:
+        report.notes.append("No persisted discovery candidates were found.")
+    if not attempts:
+        report.notes.append("No persisted scrape attempts were found.")
+    if not operational_urls and config.operational.provider is not OperationalProvider.NONE:
```
[the remark this answers is not in the document]
Comment (shaypal5) [reply src/denbust/diagnostics/discovery.py]: Gated the "No operational records were available..." note on `include_operational_matches`, so the skip path now emits only the intentional skip note instead of a misleading missing-records message. Added a test for that branch as well.

File: src/denbust/diagnostics/discovery.py [hunk 7]
```diff
@@ -0,0 +1,582 @@
...
+    candidates: list[PersistentCandidate],
+    attempts: list[ScrapeAttempt],
+    *,
+    key: str,
+) -> list[FailureSummary]:
+    candidate_by_id = {candidate.candidate_id: candidate for candidate in candidates}
+    counts: Counter[str] = Counter()
+    for attempt in attempts:
+        if attempt.fetch_status.value == "success":
+            continue
```
[the remark this answers is not in the document]
Comment (shaypal5) [reply src/denbust/diagnostics/discovery.py:578]: Updated `_build_failure_summaries()` to compare directly against `FetchStatus.SUCCESS` and imported the enum into this module. The existing failure-summary test now exercises that path.
